rust/template/d3log/src/error.rs

```diff
@@ -0,0 +1,144 @@
+// this file follows the apparently established pattern of defining a local error type and defining From
+// so as to implicity coerce all the errors thrown by '?'
+//TODOs:
+//  write a macro for the common to_string case below
+//  refactor(?) async_error so it doesn't require its users pull in a bunch of dependencies
+//  consider adding support for nesting
+//  macro for record field extraction case
+//  consider adding support for source location if thats possible
+
+use core::convert::Infallible;
+use core::fmt;
+use core::fmt::Display;
+use std::ffi::NulError;
+use std::str::Utf8Error;
+use tokio::task::JoinError;
+
+#[derive(Debug)]
+pub struct Error {
+    contents: String,
+    line: Option<u64>,
+    functionname: Option<String>,
+    filename: Option<String>,
+    parent: Box<Option<Error>>,
```

Comment:
```suggestion
    parent: Option<Box<Error>>,
```
`Box<T>` has a niche that `Option` can use, this lets us avoid heap allocations when there's no parent